README.md

```diff
@@ -106,16 +100,17 @@ autoload_rel "dir2/my_file.rb", base_dir: File.dirname(__FILE__) + "/../dir1"
 If having some problems with `autoload_all` or `autoload_rel` then set `$DEBUG=true` to see how files
 are mapped to their respective modules and classes.
 
-## Methodology (except for autoload_{all|rel})
+## Version compatibility and upgrading
+
+Prior to version 2, RequireAll attempted to automatically resolve dependencies between files, thus
+allowing them to be required in any order. Whilst convenient, the approach used (of rescuing
+`NameError`s and later retrying files that failed to load) was fundamentally unsafe and can result
+in incorrect behaviour (for example issue #8, plus more detail and discussion in #21).
 
-* Enumerate the files to be loaded
-* Try to load all of the files.  If we encounter a `NameError` loading a 
-  particular file, store that file in a "try to load it later" list.
-* If all the files loaded, great, we're done! If not, go through the
-  "try to load it later" list again rescuing `NameError` the same way.
-* If we walk the whole "try to load it later" list and it doesn't shrink
-  at all, we've encountered an unresolvable dependency.  In this case,
-  `require_all` will rethrow the first `NameError` it encountered.
+As of version 2, RequireAll will raise a `RequireAll::RequireError` if it encounters a `NameError`
+caused by a reference to an uninitialised constant during the requiring process. As such, it is not
+backwards compatible, but simple to upgrade by adding any requires to load dependencies in files
+that need them.
```

Comment:
Actually I'm not sure that I would keep this in README at all. I would move this upgrading part into CHANGES instead and introduce in README a separate paragraph about unresolved dependencies, where most of this paragraph would be. In other words - keep incompatibility and upgrading part in the CHANGES and keep unresolved dependencies part (even if it is kind of a duplicate) in README so that old require_all users can see upgrading information in the change log (I would expect to find that information in there rather in the README) and new users could see information about unresolved dependencies in the readme.

Makes sense?

Reply:
It's up to you, it's your repo! 😄

I think it's a great idea to add something to the CHANGES (which I've done), but it's quite common in READMEs for projects to mention upgrading and backwards-incompatible changes, and (personally) I think the two paragraphs make sense together, but I've moved the detail of the first paragraph to the CHANGES now.